### DISCUSSION TITLE
How to pass array to runtimeConfig by env

### DISCUSSION THREAD

**Discussion by motopods** [nuxt/nuxt#24470] [Questions]
export default {
   domains: ["111", "222"]
}
If I pass the NUXT_DOMAINS env variable, the useRuntimeConfig().domains will be overwritten to string, not an array.
NUXT_DOMAINS=111,222
May be support a seperator to split env string to an array?

**Comment by xjccc** [comment]
nuxt use `dotenv` to pass variables

`dotenv` save variables as string

you can use `JSON.parse` to parse this

**Reply by xak2000** [reply]
Where to put `JSON.parse` code though? Nuxt reads ENV variable values and sets `runtimeConfig` values **by itself**. Is there a some kind of hook that Nuxt exposes, that I'm not aware of?

Some kind of hook, that I can use to customize the process of parsing of ENV var value before Nuxt will update a `runtimeConfig` value.

**Comment by TheAlexLichter** [comment]
Defu is clever enough to recognize + parse that.
[Example](https://stackblitz.com/edit/nuxt-discussions-24470?file=nuxt.config.ts,app.vue) using `NUXT_PUBLIC_ARR=[1,2,3] nuxt dev`

**Reply by xak2000** [reply]
Ah, so, the previous @xjccc's answer is wrong. I don't need to cusomize parsing of `NUXT_*` variables. Nuxt already uses `defu` to parse them, so all I need is to have correct JSON assigned to ENV variable, right?

I tried this and it works!
```
NUXT_TRUSTED_ORIGINS=["http://localhost:8090", "http://localhost:8091"]
```
Thank you @manniL!

The tricky part is that the value needs to be a **correct JSON**. E.g. this doesn't work:
```
NUXT_TRUSTED_ORIGINS=['http://localhost:8090', 'http://localhost:8091']
```

This also doesn't work:
```
NUXT_TRUSTED_ORIGINS='http://localhost:8090', 'http://localhost:8091'
```

Unfortunately, it's not explained in the [Nuxt documentation](https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables) that Nuxt uses `defu` under the hood and you can use any valid JSON as an ENV var value to set non-string values of `runtimeConfig`. It's not that obvious, to be honest. For example, in Spring Framework, `.properties` syntax is used to read config properties from ENV vars, so `NUXT_TRUSTED_ORIGINS=http://localhost:8090,http://localhost:8091` will be the correct form.

Probably it's worth to add to the docs? With a small example of assiging a string-array (so, the reader will understand that he nees to use double-quotes, etc.).

**Reply by TheAlexLichter** [reply]
Yes, happy to see an additional paragraph with regards to `defu` + env variables 👍

**Reply by xak2000** [reply]
@manniL Do you mean I should create a PR? I'm not sure I can describe it properly as I'm not familiar with `defu` and how exatly Nuxt uses it to deserialize env vars.

**Reply by TheAlexLichter** [reply]
@xak2000 If you want to, please give it a go! Happy to help if something is unclear ☺️

**Reply by ProgramRipper** [reply]
@manniL What if I want to parse env variable as a `Uint8Array`? Maybe I should write a hook in somewhere? Is this possible?

Many cryptography lib needs bytes-like key, pass it in Base64 format by env and use as `Uint8Array` in code maybe convenient.

**Reply by TheAlexLichter** [reply]
@ProgramRipper you can convert the base64 key on the client when using the runtimeconfig to retrieve it ;)

**Reply by xak2000** [reply]
> @xak2000 If you want to, please give it a go! Happy to help if something is unclear ☺️

Unfortunately, all is unclear. :) I want to, but I'm afraid I'm not the right person to do this. I know nothing about how it really works, so how can I document it? What I really want is to make the documentation improved. If `runtimeConfig` supports JSON values (and this support is intential, not by coincidence), **it should be reflected in the docs**.

Why I can't write a PR? Not because I do not want to. I already spent more time on this problem than a PR would consume. :) But because I really do not understand many things about how this feature works, so **I don't want to lie in the docs**.

For instance:

* Is this feature intentional or just a side effect of using some 3rd party library (`defu`? I even do not know what it is, never used it. From a quick glance it is a library to merge JSON objects (apply default values conveniently) and it has nothing to do with JSON parsing)? Because, I feel like if it is not documented in the first place, maybe it is just an implementation details, and this ability (to understand env vars that contain a valid JSON) can be removed/changed in the future, so users should not use it or rely on it.

* How exactly parsing of env variables works? How can it support both strings and JSON? Does it try to parse a value as JSON and if parsing is failed, then it assumes it as string? This information is important to understand to properly document it. And this needs to be documented because as a user of Nuxt I need to understand what to expect when I'm assigning a value to `NUXT_` variable. E.g. without this information I would expect that if I write `NUXT_VAR=["foo"]` than `runtimeConfig.var` will have a `string` value `["foo"]` (literally string with brackets inside), as opposite to `string[]`. So, if I really want `runtimeConfig.var` to be a string, probably I need a workaround, like this `NUXT_VAR="[\"foo\"]"` (so, it will be parsed as JSON string).

As you can see, too many questions and misunderstanding on my side. So, it would be much more appropriate if documentation will be changed by person who really knows what they are talking about, like you @manniL. ;)

**Comment by vanling** [comment]
@TheAlexLichter Thanks for your example! Worked for al the "standard config things" 
Only not for the i18n module, seems it uses some custom .env config key handling and its doing nothing 
https://github.com/nuxt-modules/i18n/issues/3299#issuecomment-2592913651
